zebra-rpc/src/server.rs

```diff
@@ -77,7 +80,13 @@ impl RpcServer {
             // Create handler compatible with V1 and V2 RPC protocols
             let mut io: MetaIoHandler<(), _> =
                 MetaIoHandler::new(Compatibility::Both, TracingMiddleware);
-            io.extend_with(rpc_impl.to_delegate());
+            if cfg!(feature = "getblocktemplate-rpcs") {
+                #[cfg(feature = "getblocktemplate-rpcs")]
+                //io.extend_with(Rpc::to_delegate(rpc_impl));
+                io.extend_with(GetBlockTemplateRpc::to_delegate(rpc_impl));
+            } else {
+                io.extend_with(Rpc::to_delegate(rpc_impl));
+            }
```

Comment:
The issue is here. Right now (after https://github.com/ZcashFoundation/zebra/pull/5357/commits/b1a6918987ac09d44550350ef9be22511ba18a3e) the rpc server will only recognize `get_block_count` rpc mwthod when compiled with the feature while it will recognize everything except for `get_block_count` when it is not. This is expected with the current code, however:

- We should be able to call both things when the feature is enabled. Unfortunately, after the first call to `to_delegate` is made the `rpc_impl` is moved so i can't try to extend again.
OR
- A single call to `GetBlockTemplateRpc::to_delegate` should have all the rpc methods involved as it is extending what it is already there.

I don't have a solution for this yet however i will keep trying. Maybe @conradoplg , @upbqdn or @arya2 can have some idea, it seems the current design is not good enough.
-

Reply:
Another option here is to just forget about the supertrait and add the functions in the same `methods.rs` file we have now, marking each new function with the `getblocktemplate-rpcs` feature. This is a lot easier, should work and i don't see it that bad (eventually after the audit we might want to remove the rust feature for this functions anyway so it is easier to just have them located where they will be at the end).

Reply:
I think it might need a new trait just for the GBT methods? But I like your suggestion, it seems to be the simplest thing that works

Reply:
Well, i proposed that at the beginning as the easiest solution https://github.com/ZcashFoundation/zebra/issues/5305#issuecomment-1268781658 but @teor2345 want the supertrait which i think brings no big benefit (can be a bit more cleaner to have this code in separated files for some reason) but it is harder to implement.

Reply:
I'd rather we implement `GetBlockTemplateRpc` for a new struct than make it a super trait. i.e.
```suggestion
            io.extend_with(Rpc::to_delegate(rpc_impl));
            
            #[cfg(feature = "getblocktemplate-rpcs")]
            io.extend_with(GetBlockTemplateRpc::to_delegate(
                GetBlockTemplateRpcImpl::new(),
            ));
```

Reply:
So, the last approach is basically here: https://github.com/ZcashFoundation/zebra/blob/34cdcfb550590b3708f1c38fc373e6fa243e5375/zebra-rpc/src/methods.rs#L990_L1012

where there are 2 versions of the rpc implementation function, one with an error, that is for when the feature is not enabled and one with the actual implementation, for when the feature is there.